packages/venia-concept/src/components/MiniCart/cartOptions.js

```diff
@@ -0,0 +1,149 @@
+import React, { Component, Suspense } from 'react';
+import { array, func, number, shape, string } from 'prop-types';
+import { Form } from 'informed';
+import { loadingIndicator } from 'src/components/LoadingIndicator';
+import classify from 'src/classify';
+import defaultClasses from './cartOptions.css';
+import Button from 'src/components/Button';
+import Quantity from 'src/components/ProductQuantity';
+import appendOptionsToPayload from 'src/util/appendOptionsToPayload';
+
+const Options = React.lazy(() => import('../ProductOptions'));
+
+class CartOptions extends Component {
+    static propTypes = {
+        classes: shape({
+            root: string,
+            focusItem: string,
+            price: string,
+            form: string,
+            quantity: string,
+            quantityTitle: string,
+            save: string,
+            modal: string,
+            modal_active: string,
+            options: string
+        }),
+        cartItem: shape({
+            item_id: number.isRequired,
+            name: string.isRequired,
+            price: number.isRequired,
+            qty: number.isRequired
+        }),
+        configItem: shape({
+            configurable_options: array
+        }),
+        updateCart: func.isRequired,
+        hideEditPanel: func.isRequired
+    };
+
+    constructor(props) {
+        super(props);
+        this.state = {
+            optionSelections: new Map(),
+            quantity: props.cartItem.qty,
+            isLoading: false
+        };
+    }
+
+    get fallback() {
+        return loadingIndicator;
+    }
+
+    setQuantity = quantity => this.setState({ quantity });
+
+    handleSelectionChange = (optionId, selection) => {
+        this.setState(({ optionSelections }) => ({
+            optionSelections: new Map(optionSelections).set(
+                optionId,
+                Array.from(selection).pop()
+            )
+        }));
+    };
+
+    handleClick = async () => {
+        const { updateCart, hideEditPanel, cartItem, configItem } = this.props;
+        const { optionSelections, quantity } = this.state;
+        const { configurable_options } = configItem;
+        const isConfigurable = Array.isArray(configurable_options);
+        const productType = isConfigurable
+            ? 'ConfigurableProduct'
+            : 'SimpleProduct';
+
+        const payload = {
+            item: configItem,
+            productType,
+            quantity: quantity
+        };
+
+        if (productType === 'ConfigurableProduct') {
+            appendOptionsToPayload(payload, optionSelections);
+        }
+        this.setState({
+            isLoading: true
+        });
+        await updateCart(payload, cartItem.item_id);
```

Comment:
We want to avoid `await`ing redux actions in event handlers like this.

Can we refactor so that `handleClick` just updates some Redux state (via `updateCart`)?

Redux should be the source of truth that tells our app whether we're in a loading state and whether or not the edit panel should be open or closed.